peep.c

```diff
@@ -3839,6 +3839,59 @@ Perl_rpeep(pTHX_ OP *o)
                     }
                 }
             }
+            OP* rhs = cBINOPx(o)->op_first;
+            OP* lval = cBINOPx(o)->op_last;
+
+            /* Combine a simple sassign into a simple padsv lvalue
+             * into a single OP. The RHS is untouched.
+             * Removal of both sassign kids for cases like:
+             * my $foo = undef; or my $bar = $other_padsv
+             * is handled elsewhere in rpeep. */
+            if (!(o->op_private & OPpASSIGN_BACKWARDS) &&
+                !(o->op_private & OPpASSIGN_CV_TO_GV) &&
+                 lval && (lval->op_type == OP_PADSV) &&
+                 /* Note that lval could be an ex-op, e.g. if optimized
+                  * away by multideref. Don't try anything further. */
+                 (OpSIBLING(rhs) == lval)
+               ) {
+
+                /* The PADSV will typically (always?) have these flags:
+                 * SCALAR - must not be incorporated into the new op,
+                 *          which must get context from the SASSIGN
+                 * REF    - ? TODO WHAT DOES THIS EVEN MEAN HERE?
+                 * MOD    - the new OP is always for an lvalue
+                 * SPECIAL- ? TODO WHAT DOES THIS EVEN MEAN HERE?
+                 * SLABBED - probably set when the new op is created?
+                 *
+                 * Besides the op and sibling context, which carry over
+                 * the SASSIGN will have KIDS & STACKED set. */
+
+                OpTYPE_set(o, OP_PADSV_STORE);
+
+                o->op_flags = (
+                    OPf_MOD | OPf_KIDS | OPf_STACKED |
+                    (o->op_flags & OPf_WANT) |
+                    (o->op_flags & OPf_PARENS)
+                );
```

Comment:
SLABBED should be retained - it means the OP was allocated from the SLAB attached to a CV during compilation.  Clearing it may result in an invalid free() later.

Do you have examples for when REF and SPECIAL were set?

Reply:
Thanks, updated in my local copy until I fixup the ASAN errors and rebase.

I'm genuinely unsure whether to propagate the REF and SPECIAL flags or not. (`pp_sassign` doesn't use them, but I don't know if anything else would check them.)

Reply:
Now I remember, looking at it again, SLABBED is already carried over from the original _sassign_ OP because it's a separate bitfield in `BASEOP`, not part of `op_flags`.

Reply:
`(o->op_flags & OPf_WANT) | (o->op_flags & OPf_PARENS)` equals `(o->op_flags & (OPf_WANT | OPf_PARENS))`

Reply:
Merged in working copy